proposals/20240606-jwt-handling.md

```diff
@@ -0,0 +1,125 @@
+# Standardise JWTs Usage
+
+**Authors**: Kimonas Sotirchos @kimwnasptd
+
+## Requirements
+- OIDC logic, via [Istio external authoriser](https://istio.io/latest/docs/tasks/security/authorization/authz-custom/), should be adding `id_token` to http requests in `Authorization: Bearer <token>` headers
+## Scope
+This proposal aims to standardise how the Kubeflow backends should be handling the user information (name of user, groups they belong to), living in JWTs and http headers.
+
+This proposal takes as a requirement that users should be able to use K8s Tokens as `Authorization: Bearer <token>` headers in their request from inside the cluster.
+### In
+- Which component should be validating JWTs (id-tokens from OIDC or K8s ServiceAccount tokens)
+- Define where the backends should expect to find user related information
+- Define how different token "types" (dex, k8s) should be handled
+### Out
+- Proposing code changes to existing components
+## Current State
+As of Kubeflow 1.8 the user information has been injected into requests as the `kubeflow-userid` header, from the AuthService (replaced by `oauth2-proxy`). For this approach to be secure there are the following patterns that Kubeflow follows:
+1. Backends in the `kubeflow` namespace that need to know the user identity will rely on `kubeflow-userid` headers in http requests
+2. The AuthService will be adding the `kubeflow-userid` header to all authenticated requests
+3. Only requests from the Istio IngressGateway should be trusted to have the `kubeflow-userid` header
+    1. Backends that shouldn't be exposed to user namespaces (i.e. jupyter-web-app) are only reachable via the Istio IngressGateway
+    2. The KFP backend [explicitly drops requests](https://github.com/kubeflow/manifests/blob/96ce068e16b2a707464471bddc0d2a58e403d1fc/apps/pipeline/upstream/base/installs/multi-user/istio-authorization-config.yaml#L37) from user namespaces if they have this header
+    3. For in-cluster Pods that want to talk to `kubeflow` workloads, which understands identity, the Pods will need to [use a K8s ServiceAccount Token](https://www.kubeflow.org/docs/components/pipelines/v1/sdk/connect-api/#full-kubeflow-subfrom-inside-clustersub)
+### Limitations
+- Not able to express `AuthorizationPolicies` for group header in Istio
+
+The above limitation has pushed us to inject `id_tokens` to requests, via `oauth2-proxy`, to all authenticated requests. This will enable us to define policies in the future for better handling of groups.
+https://istio.io/latest/docs/tasks/security/authorization/authz-jwt/
+
+But the above creates the following issues:
+2. Ambiguity on how to handle different `id_tokens` (i.e. from Dex and from K8s)
```

Comment:
I think this is not getting rendered correctly, [check the file](https://github.com/kubeflow/manifests/blob/0121ae865418aa6b81d3e6fa8740d765972bf5e8/proposals/20240606-jwt-handling.md#limitations)

Reply:
Ah good catch, thanks!